### PR TITLE
vo_gpu: hwdec_vaapi: add dma-buf modifiers support.

### PR DESCRIPTION
**Pull request by airlied**
If the EGL extension is present, pass the modifiers for each plane
to the EGL driver.

Read this before you submit this pull request:
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Reading this link and following the rules will get your pull request reviewed
and merged faster. Nobody wants lazy pull requests.

### CODE REVIEW

**Comment by fooishbar** [comment]
Looks entirely reasonable to me, provided that the modifier is MOD_INVALID when not supplied from upstream rather than initialised to 0.

**Comment by emersion** [comment]
Seems similar to https://github.com/mpv-player/mpv/pull/9453?

File: video/out/hwdec/hwdec_vaapi_gl.c [hunk 1]
```diff
@@ -135,11 +144,18 @@ static void vaapi_gl_mapper_uninit(const struct ra_hwdec_mapper *mapper)
                         p_mapper->desc.layers[n].offset[plane]); \
             ADD_ATTRIB(EGL_DMA_BUF_PLANE ## plane ## _PITCH_EXT, \
                         p_mapper->desc.layers[n].pitch[plane]); \
+            if (p_owner->use_modifiers) {                              \
+                ADD_ATTRIB(EGL_DMA_BUF_PLANE ## plane ## _MODIFIER_LO_EXT, \
+                           p_mapper->desc.objects[p_mapper->desc.layers[n].object_index[plane]].drm_format_modifier & 0xffffffff); \
```
Comment (haasn) [inline video/out/hwdec/hwdec_vaapi_gl.c:149]: Should this be `0xfffffffflu`?

File: video/out/hwdec/hwdec_vaapi_gl.c [hunk 2]
```diff
@@ -135,11 +144,18 @@ static void vaapi_gl_mapper_uninit(const struct ra_hwdec_mapper *mapper)
                         p_mapper->desc.layers[n].offset[plane]); \
             ADD_ATTRIB(EGL_DMA_BUF_PLANE ## plane ## _PITCH_EXT, \
                         p_mapper->desc.layers[n].pitch[plane]); \
+            if (p_owner->use_modifiers) {                              \
```
Comment (haasn) [inline video/out/hwdec/hwdec_vaapi_gl.c:147]: Might as well also check that the modifier is set (i.e. not `DRM_MOD_INVALID`), just to be on the safe side. (I'm not sure where these modifiers come from but I remember running into a case similar like this when implementing this in my own code)

**Comment by haasn** [comment]
As the author of #9453 noticed, the `attribs` array could end up too small (if you happen to have 4 mapped planes - unlikely as it may be in practice).

It would be nice if the two of you could come to some resolution as to whose commit should be preferred.

**Comment by philipl** [comment]
@elfarto can you take a look. These two changes are similar but not identical. Let's make sure we don't miss any goodness from one or the other.

**Comment by elFarto** [comment]
This patch is better, save for difference hassn noticed about the attribs  array size. I had been wondering if we should maybe check that the extension exists before using it.

File: video/out/hwdec/hwdec_vaapi_gl.c [hunk 3]
```diff
@@ -135,17 +145,25 @@ static void vaapi_gl_mapper_uninit(const struct ra_hwdec_mapper *mapper)
                         p_mapper->desc.layers[n].offset[plane]); \
             ADD_ATTRIB(EGL_DMA_BUF_PLANE ## plane ## _PITCH_EXT, \
                         p_mapper->desc.layers[n].pitch[plane]); \
+            if (p_owner->use_modifiers && \
+                p_mapper->desc.objects[p_mapper->desc.layers[n].object_index[plane]].drm_format_modifier != DRM_FORMAT_MOD_INVALID) { \
+                ADD_ATTRIB(EGL_DMA_BUF_PLANE ## plane ## _MODIFIER_LO_EXT, \
+                           p_mapper->desc.objects[p_mapper->desc.layers[n].object_index[plane]].drm_format_modifier & 0xfffffffful); \
+                ADD_ATTRIB(EGL_DMA_BUF_PLANE ## plane ## _MODIFIER_HI_EXT, \
+                           p_mapper->desc.objects[p_mapper->desc.layers[n].object_index[plane]].drm_format_modifier >> 32); \
```
Comment (haasn) [inline video/out/hwdec/hwdec_vaapi_gl.c:153]: I think it would be overall more readable to just give the modifier in a name (it's inside `do` block so it won't leak outside the scope anyway). Would make the code prettier for sure. 🤷🏻‍♀️

**Comment by haasn** [comment]
LGTM

File: video/out/hwdec/hwdec_vaapi_gl.c (hunk 3 above)
Comment (airlied) [reply video/out/hwdec/hwdec_vaapi_gl.c]: done.

**Comment by orbea** [comment]
New build error:
```
../video/out/hwdec/hwdec_vaapi_gl.c:22:10: fatal error: drm_fourcc.h: No such file or directory
```
Where is this header from?

**Comment by airlied** [comment]
it's used in some other places it comes from libdrm, could just pull the one define needed from it I suppose.

**Comment by orbea** [comment]
I haven't looked at meson yet, but this is the only place that sets the right cflags which I just happened to have disabled.
```
        'name': '--drm',
        'desc': 'DRM',
        'deps': 'vt.h || consio.h',
        'func': check_pkg_config('libdrm', '>= 2.4.75'),
```
https://github.com/mpv-player/mpv/blob/a6e5eba6ab5909bbd43fe11c8cb79d7e926aedb9/wscript#L494

I suppose this just needs to do this for `vaapi` too? Or maybe can just set `drm` as a dependency?

**Comment by airlied** [comment]
I would have though drm was a vaapi dependency, it's kinda hard to follow down the meson deps list

that file is vaapi_egl
if vaapi_egl['use']
    dependencies += [vaapi_wayland['deps'], vaapi_drm['deps']]
    features += vaapi_egl['name']
    sources += files('video/out/hwdec/hwdec_vaapi_gl.c')
endif

vaapi_drm = {
    'name': 'vaapi-drm',
    'deps': dependency('libva-drm', version: '>= 1.1.0', required: get_option('vaapi-drm')),
}
vaapi_drm += {'use': vaapi['use'] and egl_drm.allowed() and vaapi_drm['deps'].found()}
if vaapi_drm['use']
    features += vaapi_drm['name']
endif

hmm maybe there is a missing drm dep in there somewhere.

**Comment by philipl** [comment]
Surprisingly, this appears to be the first actual use of libdrm includes in this vaapi hwaccel. I guess it's possible to build a useful x11 EGL build without it, so that would argue in favour of copying the define and avoiding the dependency.

**Comment by philipl** [comment]
Is this a divergence in behaviour vs waf? I'm not in a position to test until next week. @Dudemanguy

**Comment by Dudemanguy** [comment]
No, it's the same in both waf and meson. This file never had a drm dependency. It was always compiled as long at least vaapi-x11, vaapi-wayland, or vaapi-drm was available.

**Comment by airlied** [comment]
I've submitted
https://github.com/mpv-player/mpv/pull/9512

not it's biggest fan

**Comment by philipl** [comment]
I agree it's a bit ugly. The other option is to ifdef it and not set modifiers if libdrm is unavailable.

**Comment by airlied** [comment]
I think that would lead to wierd effects, since then egl and vaapi would fail to interoperate unless drm is also enabled.

**Comment by airlied** [comment]
maybe it's fine to not check the INVALID modifier here and just pass them directly from vaapi to egl
@haasn you asked for the check I think

**Comment by fooishbar** [comment]
Yeah it is. If you have `EGL_EXT_image_dma_buf_import_modifiers`, then pass along the modifier token you have for every plane - regardless of whether or not it's `INVALID`. If you do not have that EGL extension, then do not pass along any modifier token (and do not attempt modifier-aware allocation either).

**Comment by haasn** [comment]
I'm sorry, I don't remember what prompted me to add that check to the libplacebo version of this code. All I remember is that some case required it.

I could look into it if really needed, but I think the current code is fine.

**Comment by fooishbar** [comment]
Passing `INVALID` to import is not an error: it's explicitly informing the implementation that the precise layout is unknown and needs to be derived (the same as if no modifier was passed). The only erroneous case is to specify the `MODIFIER_{HI,LO}` tokens if `EGL_EXT_image_dma_buf_import_modifiers` is not supported, or to explicitly specify modifiers to the allocation path when the importer does not support the modifiers extension.

cf. [this modifier-usage doc](https://lists.freedesktop.org/archives/dri-devel/2021-September/321974.html) which is going to be merged when I've cleaned up the review comments

**Comment by haasn** [comment]
Okay, I looked into it again also:

* I currently get `INVALID` modifiers from e.g. `eglExportDMABUFImageQueryMESA`
* Passing `INVALID` is fine *iff* you have support for modifiers, so indeed only the one check would suffice
* In my own code I sometimes use `INVALID` to signal lack of modifier support, which is why I added the check on my end
* Passing `INVALID` is *equivalent* to not passing a modifier, so the current code is nonetheless correct

So my conclusion is: The check is fine but unnecessary.
